### ISSUE TITLE
n+1 / batching

### ISSUE THREAD

**Issue by zekenie** [gmac/graphql-stitching-ruby#63]
Hello!

I'm considering using this gem for a graphql-ruby / packwerk project. The idea of having in-process graphql stitching is compelling. I'm curious if you have thoughts about how I could avoid n+1 problems. I don't have a great understanding about how gateways / schema stitching tools handle this in general.

**Comment by gmac** [comment]
Hi @zekenie – sorry for the delay in reply, I missed your issue coming in. You can find the primary batching code here: https://github.com/gmac/graphql-stitching-ruby/blob/main/lib/graphql/stitching/executor.rb#L87-L130. It uses a [query batching](https://github.com/gmac/schema-stitching-handbook/wiki/Batching-Arrays-and-Queries#what-is-query-batching) strategy to combine all requests bound for the same location into a single request.

TL;DR – what this means is that stitching will make one request per location _per generation_ of data... "per generation" is key note, because there's no real optimization for getting data from A, joining that to B, then joining B's data back to A... it's a hard constraint that you're going to have to visit A twice.

Have a look at this chapter in my stitching book for best practices and to understand merging better: https://github.com/gmac/schema-stitching-handbook/tree/main/type-merging-arrays. Distilled action item would be: stitching using list queries whenever possible, and stitching will surprise you about how efficient it can be.

**Comment by gmac** [comment]
Oh, one other tip I forgot to mention that's general GraphQL but highly relevant to stitching: make sure that your root fields are batched. This anticipates multiple calls to the same root fields in a request will resolve without N+1.
